core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1357,10 +1357,12 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
   ): ZIO[R1, E2, B] =
     new ZIO.Fold[R1, E, E2, A, B](
       self,
-      ZIOFn(() => that)(_.stripFailures match {
-        case None    => that
-        case Some(c) => ZIO.halt(c)
-      }),
+      ZIOFn(() => that) { cause =>
+        cause.stripFailures match {
+          case None    => that.catchAllCause(cause2 => ZIO.halt(Cause.die(FiberFailure(cause)) ++ cause2))
```

Comment:
If we are going to wrap in `Cause.die` (I am not sure 😆 ), we could avoid `stripFailures` altogether.

Reply:
Are we? I thought `stripFailures` were used to continue execution if the error that happened previously wasn't fatal. Otherwise execution halts.

Reply:
For example `IO.die(1) orElse IO.fail(2)` would fail on the right side instead of dying on the left side.

Reply:
`stripFailures` will remove the `E` from the `fail`. But if you are taking the _whole_ cause and putting it into `Die`, then that `E` will not be visible at the type level.

We will always do `that` right after `this` fails, it's only when both fail that we take the whole left `cause`, smash it into `Die`, and then follow it by the right's (typed) failure.

You can write a test here to see the behavior of `self.catchAllCause(cause1 => that.catchAllCause(cause2 => Cause.die(FiberFailure(cause1)) ++ cause2))`

Reply:
Yes, I understand it, and if I do `self.catchAllCause(cause1 => that.catchAllCause(cause2 => Cause.die(FiberFailure(cause1)) ++ cause2))` then `ZIOSpecJvm.testOrElseDefectHandling` fails, because `(ZIO.die(ex) <> IO.unit).run` isn't equal to `Exit.die(ex)` anymore. Should I change this behaviour or does it need to be preserved?

Reply:
Hmm, what does the error change to?

Reply:
```

Reply:
```
Success(()) != Failure(Die(java.lang.Exception: Died))
Expected :Failure(Die(java.lang.Exception: Died))
Actual   :Success(())
```

Reply:
OK, right. That's because we don't want `orElse` to "recover" from defects. So if there are any defects, we fail it. Well, we can do that in another way:

```scala
       ZIOFn(() => that) { cause =>
         if (cause.died) ZIO.halt(cause)
         else that.catchAllCause(cause2 => ZIO.halt(Cause.die(FiberFailure(cause)) ++ cause2))
       }
```

This is more direct and reflects intention better than `stripFailures`.

Reply:
I agree, but `cause.died` would still recover from internal interrupts, might as well use `cause.failed` to be in line with other typed error recovery functions

Reply:
@jdegoes I'm sorry, I haven't seen your response. In my opinion, `stripFailures` is better because it returns `Option[Cause[Nothin]]`, and can pass it directly to `ZIO#halt`. Otherwise I would need to wrap `cause` into `Cause.die(FiberFailure(cause))`.

Reply:
Seems like `stripFailures` is required here to typecheck, so I think there's not much to be changed here and it's probably good to go as is, so I'll merge. 👍

Reply:
Another way would be to cast the Cause, but I guess it's not necessary tbh:

```scala
       ZIOFn(() => that) { cause =>
         if (!cause.failed) ZIO.halt(cause.asInstanceOf[Cause[E2])
         else that.catchAllCause(cause2 => ZIO.halt(Cause.die(FiberFailure(cause)) ++ cause2))
       }
```